rfcs/20221010-add-ocl-ooo-queue/README.md

```diff
@@ -0,0 +1,125 @@
+///
+/// Arguments are passed via an arguments map containing
+/// <index, memory object> pairs. The index must be one of the `DNNL_ARG_*`
+/// values such as `DNNL_ARG_SRC`, and the memory must have a memory descriptor
+/// matching the one returned by
+/// #dnnl::primitive_desc::query_md(#query::exec_arg_md, index) unless using
+/// dynamic shapes (see #DNNL_RUNTIME_DIM_VAL).
+///
+/// @param aprimitive Primitive to execute.
+/// @param astream Stream object. The stream must belong to the same engine
+///     as the primitive.
+/// @param args Arguments map.
+/// @param deps Optional vector with `cl::sycl::event` dependencies.
+///
+/// @returns Output event. The library doesn't own the event. It is the user's
+///     responsibility to manage lifetime of the event.
+inline cl_event execute(const dnnl::primitive &aprimitive,
+        const stream &astream, const std::unordered_map<int, memory> &args,
+        const std::vector<cl_event> &deps = {});
+```
+
+### Generic API
+Technically, the generic API for primitive execution is available for any stream. However, using
+such API for an out-of-order stream doesn't make much sense because a user cannot specify
+input dependencies and get an output event to embed oneDNN primitives in a larger
+dependencies graph.
+
+There is a need to specify behavior of the generic execute API for an out-of-order stream.
+There are a couple of options.
+
+#### Option 1: Prohibit Using Generic Execute API
+The idea is to check whether stream is in-order or out-of-order in the generic execute API and
+return an `invalid_arguments` status in the latter case.
+
+Pros:
+* It is clearly defined that the generic API is not suitable for out-of-order execution model
+* Less error prone implementation because there is no need to handle generic and ocl interop APIs
+differently inside the library
+
+Cons:
+* Discrepancy in behavior of SYCL and OCL runtimes. The former one allows to use the generic
+execute API
+
+#### Option 2: Allow to Use Only Generic or Only Interop Execute API
+The idea is to allow users to use both APIs but document that mixing them will result in
+undefined behavior. In order to handle dependencies between primitives submitted via the generic
+execute API (i.e. emulate in-order behavior) the stream will be used to transfer dependencies
+between execute calls. In the case of the interop API it will be the user's responsibility to
+manage dependencies.
+
+Pros:
+* User doesn't have to manage dependencies when they use only the generic API
+
+Cons:
+* Applicability is questionable. oneDNN will support both in-order and out-of-order execution
+models. If the user doesn't want to manage dependencies then they should use in-order execution
+model
+* Error prone approach. User can use both APIs for the same stream that will result in undefined
+behavior and the library cannot ensure that the mixing will not happen
+
+Proposal: go with the [option #1](#option-1-prohibit-using-generic-execute-api) because use
```

Comment:
This sounds like the library tries to be extra smart here:
- Out-of-order streams won't be the default API so I expect users to be aware of dependency tracking
- OpenCL doesn't have any guard mechanism for that

So to me it seems we should just add the interop interface and keep everything else as is.

By the way it seems like oneDNN doesn't have any checks for out-of-order queues and USM for DPC++ which is a similar use case. Is that correct?

Reply:
> This sounds like the library tries to be extra smart here

This is one way of looking at it. Another way is that we merely don't want to implement and maintain additional logic to handle such case.
In order to ensure proper synchronization between internal kernels for OOO queue we would have to request an event and pass a list of dependencies when calling `clEnqueueNDRangeKernel`. In order to do that we will use a stream to carry the dependencies passed at the interop API level and the return event that will be returned by the interop API. The list of the dependencies in the stream should be cleaned up in the end of each interop API call. But when it comes to the generic API, we would probably have to add some hook calls to clean up the dependencies in the stream.

So I think as long as we benefit from the restriction and it doesn't cause any problem it is OK to have it.

> By the way it seems like oneDNN doesn't have any checks for out-of-order queues and USM for DPC++ which is a similar use case. Is that correct?

Right, we don't. Moreover, we don't clean up list of dependencies in stream therefore we emulate in-order behavior even when queue is out-of-order and interop API is used. If we want to fix that we would also need to decide how to deal with the generic API.

Reply:
> So to me it seems we should just add the interop interface and keep everything else as is.

@echeresh this is an option. However, it assumes that the component creating the queue is the same as the component using oneDNN, which is not always true. So if we leave things as is, it will create correctness issues (e.g. calling generic execute with out-of-order queue), that could be tough to root-cause.

Reply:
Let's continue discussion [here](https://github.com/oneapi-src/oneDNN/pull/1469#discussion_r1003986659).